src/IO/RemoteReadBufferCache.cpp

```diff
@@ -0,0 +1,590 @@
+        //LOG_TRACE(&Poco::Logger::get("RemoteReadBuffer"), "from original rb {} {}", original_readbuffer->buffer().size(), original_readbuffer->offset());
+        if (status)
+            BufferBase::set(original_readbuffer->buffer().begin(), original_readbuffer->buffer().size(), original_readbuffer->offset());
+        return status;
+    }
+    return true;
+}
+
+off_t RemoteReadBuffer::seek(off_t offset, int whence)
+{
+    off_t pos_in_file = file_reader->getOffset();
+    off_t new_pos;
+    if (whence == SEEK_SET)
+        new_pos = offset;
+    else if (whence == SEEK_CUR)
+        new_pos = pos_in_file - (working_buffer.end() - pos) + offset;
+    else
+        throw Exception("expects SEEK_SET or SEEK_CUR as whence", ErrorCodes::BAD_ARGUMENTS);
+
+    /// Position is unchanged.
+    if (new_pos + (working_buffer.end() - pos) == pos_in_file)
+        return new_pos;
+
+    if (new_pos <= pos_in_file && new_pos >= pos_in_file - static_cast<off_t>(working_buffer.size()))
+    {
+        /// Position is still inside buffer.
+        pos = working_buffer.begin() + (new_pos - (pos_in_file - working_buffer.size()));
+        return new_pos;
+    }
+
+    pos = working_buffer.end();
+    auto ret_off = file_reader->seek(new_pos);
+    if (ret_off == -1)
+        throw Exception(
+            "seek file failed. " + std::to_string(pos_in_file) + "->" + std::to_string(new_pos) + "@" + std::to_string(file_reader->size())
+                + "," + std::to_string(whence) + "," + file_reader->getPath(),
+            ErrorCodes::BAD_ARGUMENTS);
+    return ret_off;
+}
+
+off_t RemoteReadBuffer::getPosition()
+{
+    return file_reader->getOffset() - available();
+}
+
+RemoteReadBufferCache::RemoteReadBufferCache() = default;
+
+RemoteReadBufferCache::~RemoteReadBufferCache() = default;
+
+RemoteReadBufferCache & RemoteReadBufferCache::instance()
+{
+    static RemoteReadBufferCache instance;
+    return instance;
+}
+
+void RemoteReadBufferCache::initOnce(const std::string & dir, size_t limit_size_)
+{
+    LOG_TRACE(log, "init local cache. path: {}, limit {}", dir, limit_size_);
+    std::lock_guard lock(mutex);
+    local_path_prefix = dir;
+    limit_size = limit_size_;
+
+    // scan local disk dir and recover the cache metas
+    std::filesystem::path root_dir(local_path_prefix);
+    if (!std::filesystem::exists(root_dir))
+    {
+        LOG_INFO(log, "{} not exists. this cache will be disable", local_path_prefix);
+        return;
+    }
+    auto callback = [this](RemoteCacheController * cntrl) { this->total_size += cntrl->size(); };
+
+    // four level dir. <schema>/<cluster>/<first 3 chars of path hash code>/<path hash code>``
+    for (auto const &schema_dir : std::filesystem::directory_iterator{root_dir})
+    {
+        for (auto const &cluster_dir : std::filesystem::directory_iterator{schema_dir.path()})
+        {
+            for (auto const &first_hash_dir : std::filesystem::directory_iterator{cluster_dir.path()})
```

Comment:
Why not include schema and cluster in the hashed part? It will get rid of this many directory traversals and will make code more generalized.

Reply:
please see the new mathod recover_cached_files_meta